# Regression Models and Complementary Datasets
**Project**: M4 Project 4
**Author**: Jacob Buysse

In this project we will be performing regression modeling on home sale pricing and them combining that dataset with a complimentary dataset to see how it might improve the model performance.

In this notebook we will be using...

In [1]:
import pandas as pd

## Part I

We need to load and clean the given dataset.  First let us load the file and look at the head/info/describe results.

In [2]:
df = pd.read_csv('./kc_house_data.csv')
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

In [4]:
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000   
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  21613.000000  21613.000000   21613.000000  21613.000000  21613.000000   
mean       7.656873   1788.390691     291.509045   1971.005136     84.402258   
std        1.175459    828.090978     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

So we have 21,613 records, 21 columns, and no missing values.

In [5]:
df.id.value_counts()

id
795000620     3
8651510380    2
2568300045    2
9353300600    2
4139480200    2
             ..
2767603649    1
1446403617    1
5602000275    1
1786200010    1
1523300157    1
Name: count, Length: 21436, dtype: int64

* `id` - This column has 21,436 distinct values (so some properties are listed multiple times; some twice and one thrice).
* `date` - This column appears to be a string in `YYYYMMDDTHHMMSS` format.
* `price` - The sale price for the house (assumed to be in USD(\\$)) ranging from \\$75,000 to \\$7,700,000 (\\$540,000 +/- \\$367,000).
* `bedrooms` - Number of bedrooms ranging from 0 to 33.
* `bathrooms` - Number of bathrooms ranging from 0 to 8 (as a float, so there must be half baths accounted for).
* `sqft_living` - Square footage of living space ranging from 290 to 13,540.
* `sqft_lot` - Lot size in square feet ranging from 520 to 1,651,000.
* `floors` - Number of floors ranging from 1 to 3.5 (not quite sure what a 1/2 floor is).
* `waterfront` - Boolean indicator (0 or 1) if the living space overlooks waterfront.
* `view` - An index from 0 (worst) to 4 (best) of how good the view is.
* `condition` - An index from 1 (worst) to 5 (best) of the condition of the living space.
* `grade` - An index from 1-3 (falls short), 4-10 (average), 11-13 (high) of the construction and design quality.
* `sqft_above` - The square footage of the house that is above ground level ranging from 290 to 9410.
* `sqft_basement` - The square footage of the house that is below ground level ranging from 0 to 4820.
* `yr_built` - The year the house was built ranging from 1900 to 2015 (note that 1900 likely means unknown).
* `yr_renovated` - The year the house was last renovated ranging from 0 to 2015 (0 likely means never renovated or unknown).
* `zipcode` - The 5-digit zip code ranging from 98001 to 98199.
* `lat` - The latitude (degrees north of equator) ranging from 47.1 to 47.7.
* `long` - The longitude (degrees east of prime meridian, negative meaning west) ranging from -122.5 to -121.3.
* `sqft_living15` - Living room size in 2015 (must assume renovations?).
* `sqft_lot15` - Lot size in 2015 (must assume renovations?).

## ID Analysis

There were duplicate property records in the original dataset.  Let us filter down to only the most recent record for each property.

In [25]:
filtered_df = df[df.groupby("id", as_index=False).date.transform(max) == df.date].copy()
filtered_df.shape

(21436, 21)

## Date Analysis

Let us extract the year the property was sold as an int so that we can derive an age property.

In [30]:
filtered_df["yr_sold"] = filtered_df.date.apply(lambda value: int(value[:4]))
filtered_df.yr_sold.describe()

count    21436.000000
mean      2014.325247
std          0.468478
min       2014.000000
25%       2014.000000
50%       2014.000000
75%       2015.000000
max       2015.000000
Name: yr_sold, dtype: float64

## Bedrooms Analysis

Let us look at the distinct values for bedrooms and a histogram plot.

In [31]:
filtered_df.bedrooms.value_counts()

bedrooms
3     9731
4     6849
2     2736
5     1586
6      265
1      194
7       38
0       13
8       13
9        6
10       3
11       1
33       1
Name: count, dtype: int64

We can see that 13 properties have a bedroom count of zero.  We can exclude these.  We can also see that there is one big outlier with a bedroom count of 33 that we should definitely exclude.  However, there are only 10 properties with more than 8 bedrooms (6 with 9, 3 with 10, and 1 with 11).  We don't have a large enough sample to split into valid train/test groups for those so we will also exclude those extremes.

We will also convert the remaining columns to a categorical type.

In [34]:
filtered_df = filtered_df[(filtered_df.bedrooms >= 1) & (filtered_df.bedrooms <= 8)]
filtered_df.bedrooms = filtered_df.bedrooms.astype('category')
filtered_df.shape

(21412, 22)